Bluetooth: controller: Initial Periodic Advertising and Sync support

cvinayak opened the pull request:
Pending:
- Generation of LE Periodic Advertising Report

cvinayak committed:
Bluetooth: controller: Add depends on BT_PER_ADV

Enable controller periodic advertising if host supports it.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add initial ULL Periodic Sync code

Add initial files and code for Periodic Sync feature support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Added bsim test to cover Periodic Sync

Added test implementation for test driven development of
Periodic Sync feature support in the controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Implementation of Periodic Sync Create

Implementation of Periodic Sync creation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add HCI integration of Periodic Sync

Added implementation to integrate the Periodic Sync LL
interface with HCI layer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add Kconfig to config Periodic Sync Sets

Added Kconfig option to configure supported simultaneous
Periodic Sync Sets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor to use ull_sync naming

Refactored to use ull_sync naming instead of ull_scan_sync.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Minor internal comments updated

Minor updates to internal comments.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Initial implementation of Periodic Sync LLL

Added initial implementation of LLL layer of Periodic Sync
creation feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Implementation of Periodic Sync Create Cancel

Implementation of Periodic Sync Create Cancel.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Added test for Periodic Sync Create Cancel

Added BabbleSim test for Periodic Sync Create Cancel.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor out clock ppm interface

Refactor out clock ppm interface so that it can be reused
for Periodic Advertising Sync feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor out common lll_abort_cb function

Refactor out common lll_abort_cb function so that it can be
used across Periodic Advertising and Periodic Advertising
Sync creation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Initial implementation of Periodic Sync setup

Added initial implementation of setting up of Periodic Sync
and scheduling the Radio Events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor out drift compensation code

Refactor out drift compensation implementation so as to
reuse it for Periodic Sync feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix minor conditional compilation

Fix minor conditional compilation around extended advertising
and scanning.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Refactor the order of file static function

Refactor the order of file static function to be consistent.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix correct CA in Periodic Advertising

Fill the correct local clock accurracy in the Sync Info
structure in the Periodic Advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix auxiliary channel scanning

Fix auxiliary channel scanning to capture the PDU end
timing that is need to correctly setup Periodic Sync.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add Periodic Sync drift compensation

Added implementation to handle Periodic Advertising clock
drift in the created Periodic Advertising Sync instance.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Added Periodic Sync Terminate and Sync Lost

Added implementation to perform Periodic Sync Terminate,
generation of Sync Lost on terminate and on remote device
termination of Periodic Advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Test Periodic Sync Create and Terminate

Added test code to exercise the Periodic Sync Creation,
Terminate and Sync Loss.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Remove cond compile on struct definition

Remove conditional compilation on structure definition in
pdu.h so as to allow inclusion of header files with
function prototypes that use those structures. I.e. to
avoid conditionally including the header file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@andrzej-kaczmarek please add yourself as reviewer to this PR (I am not able to add you).

cvinayak committed:
Bluetooth: controller: Fix missing periodic advertising start

Fix missing periodic advertising start due to missing
allocation of auxiliary context when no Extended
Advertising PDU on auxiliary channels.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix uninitialized rl_idx value

Fix uninitialized rl_idx value for the Extended Advertising
PDU received in the auxiliary channels. This caused
uninitialized rl_idx to be used by HCI layer and fail an
assert check in ull_filter.c file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Do not generate sync lost on terminate sync

Remove generation of Periodic Sync Lost HCI event generation
that is not needed as per BT Spec. v5.2 Vol.4 Part E.
Section 7.8.69.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix cond. compilation for broadcaster

Fix compilation error when building only broadcaster support
in an application.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix memory alignment violation

Fix memory word access alignment violation that causes
hardfault in nRF51x SoC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Replace use of printk with BT_WARN

Replace printk in ISR with BT_WARN to avoid problems with
co-existing with logging subsystem.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix own address type check

Fix the own address type check for whether local static
random address has been set.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Remove redundant use of update check

Remove the redundant ticker update check, ticker is not
stopped and started in Periodic Advertising unlike in a
Connection Update Procedure in connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix Periodic Advertising Context leak

Fix Periodic Advertising Context leak on Sync Lost.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add missing Periodic Adv Sync resources

Add the missing resources like ticker instances, Rx PDU
buffers and memory queue link buffers, that are required
for Periodic Advertising Sync creation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -0,0 +1,630 @@
+	int err;
+
+	err = init_reset();
+	if (err) {
```
Can we replace with 'return err;'?

cvinayak replied:
as previously explained, this is style considering future additions of code that may follow after init_reset. No action to take.

kruithofa commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2053,6 +2057,68 @@ static void le_set_ext_scan_enable(struct net_buf *buf, struct net_buf **evt)
+	ccst->status = status;
+}
+
+#if FIXME
```
We can remove the FIXME, the function ll_sync_recv_enable already returns a status code that the command is not available

cvinayak replied:
Fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -120,34 +114,226 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref, ui
+	 */
+
+	/* TODO: handle other op values */
+	if ((op != 0x03) && (op != 0x04)) {
```
We have a define for the opcodes

cvinayak replied:
Fixed

kruithofa reviewed (changes requested):
Some minor coding style comments...

cvinayak commented:
@ppryga FYI

cvinayak committed:
Bluetooth: controller: Move identical ll_scan_params_set out

Move the identical ll_scan_params_set definitions out of
extended advertising conditional compilation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -0,0 +1,301 @@
+
+	radio_aa_set(lll->access_addr);
+	radio_crc_configure(((0x5bUL) | ((0x06UL) << 8) | ((0x00UL) << 16)),
+			    (((uint32_t)lll->crc_init[2] << 16) |
```
I get use sys_get_le24. But will do in a separate PR, so that I can change other occurrences. No action in this PR.

Issue created: https://github.com/zephyrproject-rtos/zephyr/issues/27971

kruithofa replied:
Yes, a separate PR and fixing other occurrences is a good idea

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -0,0 +1,301 @@
+	radio_pkt_rx_set(node_rx->pdu);
+
+	radio_aa_set(lll->access_addr);
+	radio_crc_configure(((0x5bUL) | ((0x06UL) << 8) | ((0x00UL) << 16)),
```
Will change in a separate PR. Added in the issue https://github.com/zephyrproject-rtos/zephyr/issues/27971

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2053,6 +2057,68 @@ static void le_set_ext_scan_enable(struct net_buf *buf, struct net_buf **evt)
+	ccst->status = status;
+}
+
+#if FIXME
```
Fixed

cvinayak committed:
Bluetooth: HCI: Add the missing set per adv receive enable definition

Add the missing LE Set Periodic Advertising Receive Enable
command definition.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix set per adv recv enable HCI integration

Uncomment the code integrating the LE Set Periodic
Advertising Receive Enable command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -777,24 +784,93 @@ static void scan_recv(const struct bt_le_scan_recv_info *info,
 	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
 	printk("[DEVICE]: %s, AD evt type %u, Tx Pwr: %i, RSSI %i %s "
-	       "C:%u S:%u D:%u SR:%u E:%u Prim: %s, Secn: %s "
+	       "C:%u S:%u D:%u SR:%u E:%u I: 0x%x Prim: %s, Secn: %s "
```
Minor: Further below you print the interval as `Interval 0x%04x (%u ms)` - Would it make sense to do the same here?

cvinayak replied:
Yes, will do.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -840,44 +919,202 @@ static void test_scanx_main(void)
+	sync_create_param.sid = 0xf;
+	sync_create_param.skip = 0;
+	sync_create_param.timeout = 0xa;
+	err = bt_le_per_adv_sync_create(&sync_create_param, &sync_cb, &sync);
```
Make sure to update this to the new API (https://github.com/zephyrproject-rtos/zephyr/pull/28730) when it's merged

cvinayak replied:
Sure. Lets get yours merged first, I will rebase and test.

Thalley replied:
@cvinayak The PR has been merged

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -96,7 +96,17 @@ uint8_t ll_adv_enable(uint8_t enable);
 uint8_t ll_scan_params_set(uint8_t type, uint16_t interval, uint16_t window,
 			uint8_t own_addr_type, uint8_t filter_policy);
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+uint8_t ll_scan_enable(uint8_t enable); /* TODO: add duration and period */
```
I am working on another PR that needs a difference in the signature. https://github.com/zephyrproject-rtos/zephyr/pull/27365

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -0,0 +1,301 @@
+static int prepare_cb(struct lll_prepare_param *p);
+static void isr_rx(void *param);
+
+int lll_sync_init(void)
+{
+	int err;
+
+	err = init_reset();
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+int lll_sync_reset(void)
+{
+	int err;
+
+	err = init_reset();
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
```
This is a pattern followed so that in the future initialization specific to power and hci reset can be differentiated. The existing code is optimized out by compiler currently.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -0,0 +1,301 @@
+	LL_ASSERT(!err || err == -EINPROGRESS);
+}
+
+static int init_reset(void)
+{
+	return 0;
+}
```
This is just a template thats followed in all state/role design. For now, nothing to add. May be, I should have comments in these template functions. For now, compile optimizes these out.

cvinayak committed:
tests: Bluetooth: shell: Minor print format update

Minor print format update, updated format for Periodic
Advertising Interval, and added SID to be printed.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Minor print format update

Minor print format update, updated format for Periodic
Advertising Interval.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix SCA and Channel Map in Sync Info

Fix the population of SCA and Channel Map in Sync Info
structure.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix offset and offset units population

Fix the auxiliary pointer and sync information offset and
offset unit population. Offsets are 13-bit value, use 300 us
offset unit when offset value using 30 us offset unit does
not fit in 13-bits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

rugeGerritsen reviewed (changes requested):
Nice! I looked through the controller Kconfig configuration. I've added some suggestions on how to map the the names to the specification and how to make them more intuitive.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -206,16 +208,47 @@ uint8_t ll_sync_create_cancel(void **rx)
+	struct node_rx_pdu *node_rx;
+	memq_link_t *link_sync_lost;
 	struct ll_sync_set *sync;
+	uint32_t volatile ret_cb;
```
Why does variable need to be volatile? 

It seems that it is being cast to a `void *` (i.e. a non-volatile type) further down.

cvinayak replied:
`ull_ticker_status_give` updates ticker operation status value in `ret_cb` in ISR, and `ull_ticker_status_take` checks this status in thread context.

It is type cast to (void *) when passing to ticker API as its just used as parameter when calling the ticker operations callback. The callback in this case `ull_ticker_status_give` uses it as `uint32_t volatile *`. https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull.c#L1070

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -271,7 +271,7 @@ static void isr_rx(void *param)
 	if (crc_ok) {
 		/* TODO: */
-		printk("CRC OK\n");
+		BT_WARN("CRC OK");
```
I wonder if this is something we indeed want to have merged as is. Perhaps `BT_DBG` i more fitting, than warning us when something is OK :)

cvinayak replied:
As the feature is not complete and to avoid other debug messages, WARN is used in this PR. Shall I instead change the the message to "Periodic Event Sync-ed, report generation unsupported/work in progress." ?

Thalley replied:
`BT_WARN` just seems to be misused here :) 

If it's for debugging purposes, it should probably be `BT_DBG`. If it's for information purposes, it should probably be `BT_INFO`.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -239,7 +252,8 @@ static struct {
 						  BT_CTLR_MAX_CONN + \
-						  BT_CTLR_ADV_SET))
+						  BT_CTLR_ADV_SET + \
+						  (BT_CTLR_SCAN_SYNC_SET * 2)))
```
I'm not sure why we need `BT_CTLR_SCAN_SYNC_SET * 2` rather than just `BT_CTLR_SCAN_SYNC_SET`. Can you elaborate?

Thalley replied:
Ah, that makes sense. Yes please add a comment in the source for this, and then resolve this comment :)

Thalley reviewed (commented):
A few general comments/questions, otherwise LGTM

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -371,7 +375,7 @@ config BT_CTLR_PHY_CODED
 config BT_CTLR_CHAN_SEL_2
 	bool "Channel Selection Algorithm #2"
-	depends on (BT_CONN || BT_CTLR_ADV_PERIODIC) && BT_CTLR_CHAN_SEL_2_SUPPORT
+	depends on (BT_CONN || BT_CTLR_ADV_PERIODIC || BT_CTLR_SCAN_PERIODIC) && BT_CTLR_CHAN_SEL_2_SUPPORT
```
yes, because of BT_CONN || BT_CTLR_ADV_PERIODIC can be false.

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -45,6 +45,10 @@ config BT_CTLR_ADV_PERIODIC_SUPPORT
 	depends on BT_CTLR_ADV_EXT_SUPPORT
 	bool
 
+config BT_CTLR_SCAN_PERIODIC_SUPPORT
```
This Kconfig adds support for Periodic Advertising Sync, may be rename to `BT_CTLR_PERIODIC_SYNC_SUPPORT` or `BT_CTLR_SYNC_PERIODIC_SUPPORT` (inline with your suggestion below).

Section 4.6.13 LE Periodic Advertising is describes BIT(13) in the features supported, hence it is allowed to not have scanning for periodic advertisements and synchronization state while supporting the Scanning state.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -0,0 +1,158 @@
+{
+	/* Initialize sync pool. */
+	mem_init(ll_sync_pool, sizeof(struct ll_sync_set),
+		 sizeof(ll_sync_pool) / sizeof(struct ll_sync_set),
```
yes.

https://github.com/zephyrproject-rtos/zephyr/pull/27357/files#diff-5cf295c78bfc7caf7afcc785646fe326R55

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -0,0 +1,630 @@
+		scan_coded->per_scan.sync = NULL;
+	}
+
+	/* Check for race condition where in sync is established when sync
+	 * context was set to NULL.
+	 */
+	if (sync->timeout_reload) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
```
`sync->timeout_reload` is used to detect if a sync was established. This member is initialized to 0, but gets a valid value based on the periodic interval on sync establishment.
Thinking aloud, `scan->per_scan.sync = NULL;` should prevent any establishment race. Either the LLL ISR runs before this or after this. If run before, then sync is established and sync cancel command shall fail. Or if LLL ISR runs after, then it will not establish sync.

Ok. There is a missing NULL check.

```suggestion
	/* Check for race condition where in sync is established when sync
	 * context was set to NULL.
	 */
	if (!sync || sync->timeout_reload) {
		return BT_HCI_ERR_CMD_DISALLOWED;
	}
```

I got some new grey hair replying to justify the race condition handling ;-)

Thalley replied:
I'm not sure I understand the issue nor the solution properly yet :) 

If there is indeed another thread that may change the `sync->timeout_reload`, adding this `if` does not really fix anything as far as I can tell, since there's no way of knowing if this `if` statement will be run before or after the other thread changes that value. 

If you have a single variable that is modified or read by multiple threads, wouldn't the proper way to handle that to be adding a mutex or using an atomic block?

cvinayak replied:
Two variables are used, first `scan->per_scan.sync` is set to `NULL`, a race here is acceptable as the desired final value is `NULL`. Secondly, `sync->timeout_reload` is checked in thread mode. If ISR runs before this check, and before `scan->per_scan.sync = NULL;`, then sync will have been established and `sync->timeout_reload` will not be `0`. If the ISR runs after the check, then sync will not be established in the LLL ISR due to the `scan->per_scan.sync = NULL;`.

Thalley replied:
I see. 

So if ISR runs when `scan->per_scan.sync` isn't NULL, it'll set `sync->timeout_reload` which essentially means that the sync is established. 

So there's two common variables here, but the latter relies on the former (`scan->per_scan.sync`). 

I wonder if something like `atomic_ptr_cas` would be useful here. If both this thread and ISR use that when checking `scan->per_scan.sync`, can't we avoid the race condition?

cvinayak replied:
`atomic_ptr_cas`  uses irq_lock, that is something avoided in the controller design.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -129,16 +129,73 @@ uint8_t ll_sync_create(uint8_t options, uint8_t sid, uint8_t adv_addr_type,
+	 */
+	if (sync->is_enabled) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
```
Answered here: https://github.com/zephyrproject-rtos/zephyr/pull/27357#discussion_r498584456

cvinayak committed:
Bluetooth: controller: Rename to BT_CTLR_SYNC_PERIODIC

Rename Kconfig option BT_CTLR_SCAN_PERIODIC to
BT_CTLR_SYNC_PERIODIC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add Experimental keyword

Add Experimental keyword to LE Periodic Advertising in
Advertising and Synchronization state Kconfig title.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Use BT_INFO to log unsupported code path

Use BT_INFO to log missing support for Periodic Advertising
report generation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Added link struct allocation explanation

Added link structure allocation explanation related to HCI
event generation by the controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Fix change to Periodic Advertising API

Fix due to change to Periodic Advertising API, now requiring
callback registration.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix missing NULL check

Added a missing NULL pointer dereference check and made the
structure member scan->per_scan.sync and timeout_reload as
volatile, as they are modified in ISR context.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (approved):
LGTM

kruithofa reviewed (approved):
LGTM

rugeGerritsen reviewed (approved):
I've only reviewed the new Kconfig symbols.

cvinayak commented:
DNM, would like rebase over https://github.com/zephyrproject-rtos/zephyr/pull/28950
